rfcs/0038-unified-device-target-and-memory-scope-planning.md

```diff
@@ -0,0 +1,244 @@
+   | --------------------- | -------------------------- | ----------------------------------------------- |
+   | Source of annotations | `on_device`, `device_copy` | `FTVMAnnotateTarget`, `MergeComposite`+patterns |
+   | Target of planning    | DLDeviceType               | Toolchain name                                  |
+   | Propagation           | Unification based          | Ad-hoc                                          |
+   | Relay support         | Full                       | First-order, no ADTs                            |
+   | Delimiting            | insert `device_copy`       | insert `compiler_begin`, `compiler_end`         |
+   | Multiple per expr     | No                         | Yes (though always picks first)                 |
+   | Hoists into functions | No                         | Yes                                             |
+   | Customized heuristics | No                         | No                                              |
+
+   Taking the 'upper bound' of the two implementations seems ideal, especially to address issues 4 (limitation
+   of device planning) and 5 (limitation of BYOC) above.
+
+Our proposal is:
+1. We introduce a new FFI-friendly class to represent a *S*torage or *E*xecution *Scope*:
+
+   ```
+   class SEScope {
+     DLDeviceType device_type;
+     int virtual_device_id;
+     Target target;
+     String memory_scope;
+   }
+   ```
+
+   We allow each of these fields to be independently 'constrained' (ie have a specific value) or
+   'unconstrained' (no specific value for the field is known yet). In particular, it is valid for
+   an `SEScope` to contain only a `device_type`. However if the `target` field is defined then
+   `device_type` must equal `target->kind->device_type`.
+
+2. At this stage we leave the `memory_scope` field uninterpreted. For example, we don't attempt to
+   represent that, eg, `"global"` on a `kDLCPU` is the same memory area as `"host"` on a `kDLCUDA` and thus no
+   `device_copy` operation is required between those scopes. We'll pick this issue up again after
+   [RFC #9](https://github.com/apache/tvm-rfcs/blob/main/rfcs/0009_Unified_Static_Memory_Planning.md)
+   has landed.
+
+3. The `on_device` and `device_copy` call attributes use `SEScope`s instead of integers. However the Python
+   bindings for these 'operators' continue to accept a `Device` for convenience. The machinery in `LowerTEPass`
+   which resolves `DLDeviceTypes` to `Targets` is moved up in the compilation flow and becomes part of
+   `PlanDevices`. In particular, any `SEScope` encountered during device planning is 'canonicalized' to fill
+   in a `Target` by the same lookup as we do today. This means we continue to support the easy shorthand of
+   referring to devices by the `DLDeviceType` alone. However, advanced users can supply a `SEScope` to these
+   operators which contains the exact `Target` to use.
```

Comment:
what would be roughly the deprecation plan here? eventually we ban all the inputs to the compiler which could refer to SEScope in terms of DLDeviceType and then tighten the typing requirements here? this would be a backwards-incompatible Relay change. cc @jroesch